token-lending/program/src/processor.rs

```diff
@@ -1714,6 +1720,68 @@ fn unpack_mint(data: &[u8]) -> Result<Mint, LendingError> {
     Mint::unpack(data).map_err(|_| LendingError::InvalidTokenMint)
 }
 
+fn get_pyth_product_attribute<'a, T>(ite: &mut T) -> String
+where
+    T: Iterator<Item = &'a u8>,
+{
+    let mut len = *ite.next().unwrap() as usize;
+    let mut val = String::with_capacity(len);
+    while len > 0 {
+        val.push(*ite.next().unwrap() as char);
+        len -= 1;
+    }
+    val
+}
+
+fn get_pyth_price(
+    pyth_price_account_info: &AccountInfo,
+    clock: &Clock,
+) -> Result<u64, ProgramError> {
+    let pyth_price_data = &pyth_price_account_info.try_borrow_data()?;
+    let pyth_price = pyth_client::cast::<pyth_client::Price>(pyth_price_data);
+
+    match pyth_price.ptype {
+        pyth_client::PriceType::Price => {}
+        _ => {
+            msg!("Oracle price type is invalid");
+            return Err(LendingError::InvalidOracleConfig.into());
+        }
+    }
+    if pyth_price.valid_slot < clock.slot {
+        msg!("Oracle price is stale");
+        return Err(LendingError::InvalidOracleConfig.into());
+    }
+
+    let price: u64 = pyth_price
+        .agg
+        .price
+        .checked_abs()
+        .ok_or(LendingError::MathOverflow)?
+        .try_into()
+        .map_err(|_| LendingError::MathOverflow)?;
+
+    let pyth_exponent = pyth_price
+        .expo
+        .checked_abs()
+        .ok_or(LendingError::MathOverflow)?
+        .try_into()
+        .map_err(|_| LendingError::MathOverflow)?;
+
+    let pyth_decimals = 10u64
+        .checked_pow(pyth_exponent)
+        .ok_or(LendingError::MathOverflow)?;
+
+    // @FIXME: magic number
+    let quote_decimals = 10u64.pow(6);
```

Comment:
This might be too small, imagine a token with a really high circulating supply which is worth a really small amount. Maybe we can store both the price and the exponent and then determine market value directly from those?

Reply:
Fixed.